pkg/asset/machines/clusterapi.go

```diff
@@ -227,24 +227,16 @@ func (c *ClusterAPI) Generate(ctx context.Context, dependencies asset.Parents) e
 		}
 		client := icazure.NewClient(session)
 
-		if len(mpool.Zones) == 0 {
+		azs, err := client.GetAvailabilityZones(ctx, ic.Platform.Azure.Region, mpool.InstanceType)
+		if err != nil {
+			return fmt.Errorf("failed to fetch availability zones: %w", err)
+		}
+		mpool.Zones = azs
```

Comment:
Isn't this going to overwrite zones explicitly set in the install-config.yaml?

Reply:
I think the problem with the code before is that if the user hadn't set any zones in the install-config, then we'd set `mpool.Zones = []string{""}`. That means the next check for `if len(mpool.Zones) == 0` would fail, since `len(mpool.Zones) == 1`, so we wouldn't discover the zones available for the region before generating the machines.

Reply:
> Isn't this going to overwrite zones explicitly set in the install-config.yaml?

Yes. I didn't realize that you could manually set zones in the install-config. I'll update this.

Reply:
> I think the problem with the code before is that if the user hadn't set any zones in the install-config, then we'd set `mpool.Zones = []string{""}`. That means the next check for `if len(mpool.Zones) == 0` would fail, since `len(mpool.Zones) == 1`, so we wouldn't discover the zones available for the region before generating the machines.

Yup, that is indeed what was happening.

Reply:
Fixed.